src/PlatformEngines.jl

```diff
@@ -648,34 +661,22 @@ function unpack(tarball_path::AbstractString, dest::AbstractString;
                 verbose::Bool = false)
     # The user can force usage of our dereferencing workarounds for filesystems
     # that don't support symlinks, but it is also autodetected.
-    copyderef = get(ENV, "BINARYPROVIDER_COPYDEREF", "") == "true" ||
-                (tempdir_symlink_creation && !probe_symlink_creation(dest))
-    copyderef2 = get(ENV, "BINARYPROVIDER_COPYDEREF2", "") == "true" ||
-                (!copyderef && !tempdir_symlink_creation && !probe_symlink_creation(dest))
-    # If we should "copyderef" what we do is to unpack into a temporary directory,
-    # then copy without preserving symlinks into the destination directory.  This
-    # is to work around filesystems that are mounted (such as SMBFS filesystems)
-    # that do not support symlinks.  Note that this does not work if you are on
-    # a system that completely disallows symlinks (Even within temporary
-    # directories) such as Windows XP/7.
-    # The second workaround now also works for filesystems without symlinks. Here
-    # we exclude symlinks from being exctracted and copy the sources of the symlinked
-    # files instead.
-    true_dest = dest
-    if copyderef
-        dest = mktempdir()
-    end
+    copyderef = (get(ENV, "BINARYPROVIDER_COPYDEREF", "") == "true") || !probe_symlink_creation(dest)
+
+    # If we should "copyderef" what we do is to unpack everything except symlinks
+    # then copy the sources of the symlinks to the destination of the symlink instead.
+    # This is to work around filesystems that are mounted (such as SMBFS filesystems)
+    # that do not support symlinks.
 
-    # unpack into dest
-    mkpath(dest)
     excludelist = ""
     symlinks = []
 
-    if copyderef2
+    if copyderef
         symlinks = list_tarball_symlinks(tarball_path)
         if length(symlinks) > 0
-            excludelist = joinpath(dest, "excludelist.txt")
-            write(excludelist, join([s[1] for s in symlinks], "\n"))
+            (excludelist, io) = mktemp()
+            write(io, join([s[1] for s in symlinks], "\n"))
```

Comment:
Same thine here, use

```julia
mktemp() do (path, io)
	write(io, join([s[1] for s in symlinks], "\n"))
end
```

It's more fault-tolerant as it will catch exceptions that happen within `write()` and still close `io` when it breaks out.

Reply:
never used `do ... end` blocks until now, but it returns an error
```
MethodError: no method matching (::getfield(Main, Symbol("##19#21")))(::String, ::IOStream)
Closest candidates are:
mktemp(::getfield(Main, Symbol("##19#21")), ::String) at file.jl:555
mktemp(::Function) at file.jl:553
top-level scope at none:0
```

Reply:
I think it needs to be:
```Julia
mktemp() do path, io
	write(io, join([s[1] for s in symlinks], "\n"))
end
```
Since `(path, io)` does destructuring, which turns the function into a one arg function!

Reply:
I tested your proposal, but I need `tmpfile` later for `rm(tmpfile; force = true)`
So the version would look like
```
local tmpfile
mktemp() do path, io
	tmpfile = path
	write(io, "Demo file for tar listing (Julia package BinaryProvider.jl)")
	close(io)
end
```
Should I proceed with this version?

Reply:
I think the whole idea of the closure form is, that it deletes everything after the function is executed:
```
  mktemp(f::Function, parent=tempdir())

  Apply the function f to the result of mktemp(parent) and remove the
  temporary file upon completion.
```

Reply:
same goes for the `io`, so no need to close it!

Reply:
but I have somthing like

```
if copyderef  <create excludelist file> end
<unpack exempting exludelist>
if copyderef  <remove excludelist file> end
```
So I don't see how to use the do block.
Maybe, I am overlooking something?

Reply:
Maybe it's not the right choice here then ;)

Reply:
Simon is right; this is not the right choice in this case. :)  Sorry, I didn't fully realize all the places this was being used.  :)